chore(cli): add Execute to task delete.

bvtujo opened the pull request:
<!-- Provide summary of changes -->
Adds Execute logic to the new 'task delete' command. 

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, 77" -->

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

bvtujo committed:
chore(cli): add execute method to task delete

bvtujo committed:
chore(cli): add Execute and related methods for task delete

bvtujo committed:
chore(ecs): fix bugs in task stopping code

bvtujo committed:
chore(cecs): add abstraction for stopping appenv and default tasks

bvtujo committed:
chore(cli): delete ECR images using default session, not env manager

bvtujo committed:
chore(cfn): delete task stacks with role arn only when they're in an env

bvtujo committed:
chore(ecs): add tests for StopTask abstractions

bvtujo committed:
chore(cli): use new ecs abstraction in job delete

bvtujo committed:
chore(cfn): add tests for task delete execution

bvtujo committed:
Merge branch 'mainline' into task-delete-execute

iamhopaul123 commented on internal/pkg/cli/task_delete.go:
```diff
@@ -262,26 +281,98 @@ func (o *deleteTaskOpts) askTaskName() error {
 
 func (o *deleteTaskOpts) Execute() error {
 	// Get clients.
+	sess, err := o.getSession()
```
Can we cache the session? For example make `o.sess` to be a session and
```go
func (o *deleteTaskOpts) getSession() (*awssession.Session, error) {
   if o.sess != nil {
      return o.sess, nil
   }
   ...

```

bvtujo replied:
Good idea. I've cached the session and also the task information, which was previously getting called twice.

iamhopaul123 commented on internal/pkg/cli/task_delete.go:
```diff
@@ -262,26 +281,98 @@ func (o *deleteTaskOpts) askTaskName() error {
+	taskDeleter := o.newTaskDeleter(sess)
+
+	// ECR Deletion happens from the default profile in app delete. We can do it here too.
+	defaultSess, err := o.sess.DefaultWithRegion(aws.StringValue(sess.Config.Region))
```
Could `sess.Config.Region` be `nil` if it is using default cluster instead of app env? If so, does `DefaultWithRegion("")` work?

bvtujo replied:
Good catch!

iamhopaul123 commented on internal/pkg/cli/task_delete.go:
```diff
@@ -262,26 +281,98 @@ func (o *deleteTaskOpts) askTaskName() error {
+	}
+
+	ecsClient := o.newTaskListerStopper(sess)
+	taskDeleter := o.newTaskDeleter(sess)
```
Maybe init those at the nearest place where they are firstly used.

bvtujo replied:
There's been a big rework of the execute logic; I think it reads a little better now.

iamhopaul123 commented on internal/pkg/cli/interfaces.go:
```diff
@@ -348,6 +349,8 @@ type appResourcesGetter interface {
 
 type taskDeployer interface {
```
Since we add two more methods, what do you think renaming `taskDeployer` to be maybe `taskDeployManager`?

efekarakus replied:
:+1: Or creating a separate interface if they don't need to be grouped all three together

bvtujo replied:
done; split into `taskDeployer` and `taskStackManager`

iamhopaul123 commented on internal/pkg/cli/task_delete.go:
```diff
@@ -262,26 +281,98 @@ func (o *deleteTaskOpts) askTaskName() error {
+
+	// Get information about the task stack. This struct will be used to get the names of the ECR
+	// repo and task stack.
+	taskInfo, err := taskDeleter.GetTaskStack(o.name)
```
`taskDeleter.GetTaskStack` looks weird. Maybe we should consider renaming `taskDeleter` to be `taskStackManager`.

iamhopaul123 commented on internal/pkg/cli/task_delete.go:
```diff
@@ -262,26 +281,98 @@ func (o *deleteTaskOpts) askTaskName() error {
+	taskInfo, err := taskDeleter.GetTaskStack(o.name)
+	if err != nil {
+		return fmt.Errorf("retrieve stack information for task %s: %w", o.name, err)
+	}
```
We won't use `taskInfo` until https://github.com/aws/copilot-cli/pull/1840/files#diff-9a3336010abc0c4ff97fbe6b42f8da69a79565fb6065075d7dbb8153bd3e4250R363. What do you think moving this part below?

iamhopaul123 commented on internal/pkg/cli/task_delete.go:
```diff
@@ -262,26 +281,98 @@ func (o *deleteTaskOpts) askTaskName() error {
+	}
 
+	// Clear repository.
+	err = ecrDeleter.ClearRepository(taskInfo.ECRRepoName())
```
```suggestion
	err = o.newImageRemover(defaultSess).ClearRepository(taskInfo.ECRRepoName())
```

iamhopaul123 commented on internal/pkg/cli/task_delete.go:
```diff
@@ -262,26 +281,98 @@ func (o *deleteTaskOpts) askTaskName() error {
 	return nil
 }
 
 func (o *deleteTaskOpts) Execute() error {
```
I like the previous layout of `Execute` function. What do you think following the previous stub function, without adding distracting comments so that the workflow can be clearer?
```go
func (o *deleteTaskOpts) Execute() error {
	// Get clients.
        sess, err := o.getSession()
	// Stop tasks.
        o.stopTasks()
	// Clear repository.
        o.clearRepo()
	// Delete stack.
        o.deleteStack()
	return nil
}
```

iamhopaul123 commented on internal/pkg/cli/task_delete.go:
```diff
@@ -262,26 +281,98 @@ func (o *deleteTaskOpts) askTaskName() error {
+	// Get running tasks in family.
+	var tasks []*awsecs.Task
+	if o.defaultCluster {
+		tasks, err = ecsClient.ListActiveDefaultClusterTasks(ecs.ListTasksFilter{
+			TaskGroup: o.name,
+		})
+		if err != nil {
+			o.spinner.Stop(log.Serrorln("Error listing running tasks."))
+			return fmt.Errorf("list running tasks in default cluster: %w", err)
+		}
```
I feel `StopDefaultClusterTasks` doesn't abstract enough logic. We probably don't need to list active task here but embed the logic in `StopDefaultClusterTasks` so that `StopDefaultClusterTasks` takes the filter:

```suggestion
		tasks, err = ecsClient.StopDefaultClusterTasks(ecs.StopTasksFilter{
			TaskGroup: o.name,
		})
		if err != nil {
			o.spinner.Stop(log.Serrorln("Error stopping running tasks."))
			return fmt.Errorf("stop running tasks in default cluster: %w", err)
		}
```
What do you think?

bvtujo replied:
I've gone even further than this and embedded the listing logic in several new top level functions in the `ecs` package: 
* `StopDefaultClusterTasks(familyName string)` which stops one-off copilot tasks in the default cluster.
* `StopAppEnvOneOffTasks(app, env, family string)` which stops one-off copilot tasks in the given environment.
* `StopActiveWorkloadTasks(app, env, workload string)` which stops all copilot tasks from the given workload and env. 
These abstractions are easier to work with and mock, and I've added them in `job delete` and `task delete` where applicable. 

The `List` methods are still used by the `selector` package but are not needed in `cli`, so I've removed them from `interfaces.go`.

bvtujo committed:
chore(ecs): refactor API for higher abstraction when stopping tasks

This commit refactors the API of the ecs package to allow the following
top level functions:
* StopDefaultClusterTasks
* StopAppEnvOneOffTasks
* StopActiveWorkloadTasks

All of these operate only on copilot tasks, whose families are of the form
of either `copilot-$task` or `$app-$env-$wkld`. These functions abstract the
logic of listing tasks in either the default or environment cluster and
stopping all instances.

efekarakus commented on internal/pkg/cli/interfaces.go:
```diff
@@ -495,8 +498,11 @@ type activeWorkloadTasksLister interface {
 
-type tasksStopper interface {
-	StopTasks(tasks []string, opts ...ecs.StopTasksOpts) error
+type tasksListerStopper interface {
```
nit: the convention is `tasksListStopper` (example [`io.ReadWriteCloser`](https://golang.org/pkg/io/#ReadWriteCloser))

efekarakus commented on internal/pkg/cli/task_delete.go:
```diff
@@ -7,12 +7,17 @@ import (
 	"fmt"
 
+	"github.com/aws/aws-sdk-go/aws"
 	awssession "github.com/aws/aws-sdk-go/aws/session"
```
nit: can we get rid of the `awssession` rename? I can't see other session pkgs 😲

bvtujo committed:
chore(cli): remove unused interface

efekarakus commented on internal/pkg/cli/task_delete.go:
```diff
@@ -262,26 +290,131 @@ func (o *deleteTaskOpts) askTaskName() error {
-	// Get clients.
-
-	// Stop tasks.
+	// Stop all running tasks in given family.
```
nice! I love this new revision with moving individual sections to separate methods, a lot easier to follow. We can even get rid of these comments now.

iamhopaul123 replied:
Awesome! Yeah agree since now the code itself should be already descriptive enough.

bvtujo committed:
chore(cli): rename interface taskDeleter to taskDeployManager

efekarakus commented on internal/pkg/ecs/ecs.go:
```diff
@@ -147,8 +150,9 @@ type ListActiveAppEnvTasksOpts struct {
-	TaskID    string
+	TaskGroup  string // Returns only tasks with the given TaskGroup name.
+	TaskID     string // Returns only tasks with the given ID.
+	OneOffOnly bool   // Returns only tasks with the `copilot-task` tag.
```
What do you think of:
```suggestion
	CopilotOnly bool   // Returns only tasks with the `copilot-task` tag.
```

efekarakus commented on internal/pkg/ecs/ecs.go:
```diff
@@ -180,13 +184,64 @@ func (c Client) ListActiveDefaultClusterTasks(filter ListTasksFilter) ([]*ecs.Ta
+}
+
+// StopAppEnvOneOffTasks stops all one-off tasks in the given application and environment with the family name.
+func (c Client) StopAppEnvOneOffTasks(app, env, family string) error {
```
I think we can just drop `AppEnv` from all the method names, this pkg is an abstraction on top of `internal/aws/ecs` for Copilot specific usecases. I feel like it would read nicer what do you think?

Maybe we can do `StopOneOffTasks`?

efekarakus commented on internal/pkg/ecs/ecs.go:
```diff
@@ -180,13 +184,64 @@ func (c Client) ListActiveDefaultClusterTasks(filter ListTasksFilter) ([]*ecs.Ta
 }
 
+// StopActiveWorkloadTasks stops all tasks in the given application, enviornment, and workload.
+func (c Client) StopActiveWorkloadTasks(app, env, workload string) error {
```
nit: can we remove "Active"

efekarakus commented on internal/pkg/ecs/ecs.go:
```diff
@@ -180,13 +184,64 @@ func (c Client) ListActiveDefaultClusterTasks(filter ListTasksFilter) ([]*ecs.Ta
+}
+
+// StopAppEnvTasks stops all tasks in the given application and environment in the given family.
+func (c Client) stopAppEnvTasks(app, env, family string, oneOffOnly bool) error {
```
The bool is not used, can we remove it?

efekarakus commented on internal/pkg/ecs/ecs.go:
```diff
@@ -196,21 +251,26 @@ func (c Client) listActiveCopilotTasks(opts listActiveCopilotTasksOpts) ([]*ecs.
 }
 
-func filterCopilotTask(tasks []*ecs.Task, taskID string) []*ecs.Task {
+func filterCopilotTask(tasks []*ecs.Task, taskID string, oneOffOnly bool) []*ecs.Task {
```
nit: The recommendation is to create multiple functions instead of using flag arguments (https://martinfowler.com/bliki/FlagArgument.html) so for example we can do:

```go
func filterTasksByID(tasks []*ecs.Task, taskID string) []*ecs.Task { ... }

func filterCopilotTasks(tasks []*ecs.Task, taskID string) []*ecs.Task {
   var filteredTasks []*ecs.Task
   for _, t  := range filterTasksByID(tasks, taskID) {
      // loop through the tags.
   }
   ...
}
```

efekarakus commented on internal/pkg/cli/task_delete.go:
```diff
@@ -7,12 +7,17 @@ import (
 	"github.com/aws/copilot-cli/internal/pkg/config"
+	"github.com/aws/copilot-cli/internal/pkg/deploy"
 	"github.com/aws/copilot-cli/internal/pkg/deploy/cloudformation"
+	"github.com/aws/copilot-cli/internal/pkg/ecs"
```
nice! I really appreciate how between the first and second revision the dependency on `aws-sdk-go/ecs` went away. As we work on https://github.com/aws/copilot-cli/issues/1852 it will be a lot easier to migrate to v2 if all sdk structs are contained within `internal/aws`

iamhopaul123 commented on internal/pkg/cli/task_delete.go:
```diff
@@ -262,26 +290,131 @@ func (o *deleteTaskOpts) askTaskName() error {
+	var defaultSess *awssession.Session
+	var err error
+	if o.defaultCluster {
+		defaultSess, err = o.sess.Default()
```
Is `o.sess.Default()` equal to `o.getSession()` when `o.defaultCluster` is true?

bvtujo replied:
yes, and I've made this change.

iamhopaul123 commented on internal/pkg/cli/task_delete.go:
```diff
@@ -262,26 +290,131 @@ func (o *deleteTaskOpts) askTaskName() error {
+	// a default session in whichever region we're deleting from.
+	var defaultSess *awssession.Session
+	var err error
+	if o.defaultCluster {
+		defaultSess, err = o.sess.Default()
+		if err != nil {
+			return fmt.Errorf("get default session for ECR deletion: %s", err)
+		}
+	} else {
+		regionalSession, err := o.getSession()
+		if err != nil {
+			return err
+		}
+		defaultSess, err = o.sess.DefaultWithRegion(aws.StringValue(regionalSession.Config.Region))
+		if err != nil {
+			return fmt.Errorf("get default session for ECR deletion: %s", err)
+		}
+	}
```
Maybe
```suggestion
	defaultSess, err = o.getSession()
	if err != nil {
		return err
	}
	if !o.defaultCluster {
		defaultSess, err = o.sess.DefaultWithRegion(aws.StringValue(regionalSession.Config.Region))
		if err != nil {
			return fmt.Errorf("get default session for ECR deletion: %s", err)
		}
	}
```

bvtujo committed:
chore(cli): split taskDeployManager into taskDeployer and taskStackManager

bvtujo committed:
chore(cli): don't rename session to awssession

iamhopaul123 commented on internal/pkg/cli/task_delete.go:
```diff
@@ -262,26 +290,131 @@ func (o *deleteTaskOpts) askTaskName() error {
+		return err
+	}
+	o.spinner.Start(fmt.Sprintf("Emptying ECR repository for task %s.", color.HighlightUserInput(o.name)))
+	err = o.newImageRemover(defaultSess).ClearRepository(info.ECRRepoName())
```
I was wondering if we need `GetTaskStack`: looks like the reason why we get the task stack is to get the ECR repo name, which is simply `"copilot"+o.name`. Can we make best efforts trying to empty the ECR repo?

Also I think we should try to make sure `copilot task delete` is idempotent. Users shouldn't get interrupted if any of the resources have already been deleted before running the command. What do you think?

bvtujo replied:
We keep the TaskStackInfo around because that struct is the input to `DeleteTask` and it's convenient to use to grab the ECR Repo name. Since we cache it we only have to call the underlying method once.

iamhopaul123 replied:
Oh got it. Saw how the info gets cached. But looks like it will error out if the stack doesn't exist? Maybe we should format the ECR repo name here for repo cleaning, and do `GetTaskStack` when we are trying to delete the stack. And skip deleting if the stack doesn't exist. What do you think?

bvtujo replied:
Okay, I've added some idempotency checking for the error which gets returned when the stack doesn't exist and added a unit test for that case.

bvtujo committed:
chore(ecs): rename functions and refactor filter

bvtujo committed:
chore(cli): streamline default session for ecr emptying

bvtujo committed:
chore(cli): make task delete idempotent

efekarakus commented on internal/pkg/ecs/ecs.go:
```diff
@@ -180,13 +184,65 @@ func (c Client) ListActiveDefaultClusterTasks(filter ListTasksFilter) ([]*ecs.Ta
+}
+
+// stopTasks stops all tasks in the given application and environment in the given family.
+func (c Client) stopTasks(app, env, family string, oneOffOnly bool) error {
```
nit: 
```suggestion
func (c Client) stopTasks(app, env, filter ListTasksFilter) error {
```
and this way the caller can do:
```go
return c.stopTasks(app, env, ListTasksFilter{
    TaskGroup: fmt.Sprintf(fmtWorkloadTaskDefinitionFamily, app, env, workload),
})
// or
c.stopTasks(app, env,  ListTasksFilter{
    TaskGroup:fmt.Sprintf(fmtTaskTaskDefinitionFamily, family),
    CopilotOnly: true,
})

```

efekarakus reviewed (approved):
🚀

iamhopaul123 commented on internal/pkg/cli/task_delete.go:
```diff
@@ -262,26 +290,130 @@ func (o *deleteTaskOpts) askTaskName() error {
+	info, err := o.newStackManager(sess).GetTaskStack(o.name)
+
+	if err != nil {
+		// If the stack doesn't exist, return nil.x
```
nit
```suggestion
		// If the stack doesn't exist, return nil.
```

bvtujo committed:
chore(cli): add validation for fully qualified task names

bvtujo committed:
chore(ecs): refactor stopTasks api

iamhopaul123 reviewed (approved):
ship it 🚢

mergify[bot] committed:
Merge branch 'mainline' into task-delete-execute
